[fx] add support for custom tracer in __reduce_package__

bradleyhd opened the pull request:
Summary: Adds support for specifying a custom Tracer to use when deserializing a GraphModule from a torch package, similar to `__reduce_deploy__`.

Differential Revision: D30019214

ansley reviewed (approved):
LGTM, but I would get @suo to look over it as well

suo commented on torch/fx/graph_module.py:
```diff
@@ -546,7 +547,7 @@ def __reduce_deploy__(self, importer: Importer, tracer_cls: Type = None):
         return (reduce_deploy_graph_module, (dict_without_graph, import_block, tracer_cls))
 
-    def __reduce_package__(self, exporter: PackageExporter):
+    def __reduce_package__(self, exporter: PackageExporter, tracer_cls: Type = None):
```
Hmm, I should have said this on the reduce_deploy PR but seeing this again I realize: my preference is to add `tracer_cls` to the state of GraphModule rather than thread it through the reduce protocol as we are doing here. Two reasons for this:
1. The reduction function arguments are BC sensitive, so I'd prefer not to add new ones (to be clear, this is a worthy use case for it, but if there's an alternative maybe we should do that).
2. The current implementation requires that an inherited GraphModule class be defined so that we can curry the arguments together (in this case it is `DperGraphModule`) . If we make it automatically the case that all GraphModules store their tracer cls, then all users would benefit without having to do anything I think.

What do you think @bradleyhd @HarutMov?

HarutMov replied:
Looks like this approach was already discussed here : https://github.com/pytorch/pytorch/issues/54573. At that time it was not possible to pickle Tracer object, we need to check whether it is still a blocker.

suo replied:
we're not pickling the tracer object though, just the class, which should be fine to pickle

HarutMov commented on torch/fx/_symbolic_trace.py:
```diff
@@ -532,7 +532,8 @@ def trace(self, root: Union[torch.nn.Module, Callable], concrete_args: Optional[
             fn = root
-        self.graph = Graph()
+
+        self.graph = Graph(tracer_cls=self.__class__)
```
You may want to wrap self.__class__ with Optional to silence mypy warning: https://github.com/pytorch/pytorch/runs/3223778148

suo reviewed (commented):
This lgtm, but can we add a test on the OSS side :)

bradleyhd commented:
> This lgtm, but can we add a test on the OSS side :)

updated with a test!

suo reviewed (approved):
wooo!

bradleyhd committed:
[fx] store Tracer class on Graph and GraphModule for package deserialization (#62497)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/62497

Previously named: add support for custom tracer in __reduce_package__

Stores a Tracer class on a Graph created by Tracer, and copies the Tracer class into the GraphModule's state so that when a GraphModule is packaged by torch package, it can be reconstructed with the same Tracer and GraphModule class name.

Reviewed By: suo

Differential Revision: D30019214

fbshipit-source-id: 3dda540e9bc49a51ac258a31b47ba3e542deb79b
